Improve docs

RemDelaporteMathurin opened the pull request:
## Proposed changes

This PR has been a long time coming... Flushing out the User Guide of the documentation.

## Types of changes

What types of changes does your code introduce to FESTIM?
<!--Put an `x` in the boxes that apply-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [x] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

<!--Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.-->

- [ ] Black formatted
- [ ] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

KulaginVladimir commented:
Looks nice. Several minor suggestions. @RemDelaporteMathurin, what do you think of adding some information about:
- The Soret coefficient in **Materials**
- The stepsize settings (adaptive stepsize)
- An example of the `PointValue` usage in **PP**

RemDelaporteMathurin commented:
> Looks nice. Several minor suggestions. @RemDelaporteMathurin, what do you think of adding some information about:
> 
> * The Soret coefficient in **Materials**
> * The stepsize settings (adaptive stepsize)
> * An example of the `PointValue` usage in **PP**

Done! let me know what you think.

KulaginVladimir commented:
Perfect! Though I didn't find an example of the `InitialCondition` usage (nowhere). Maybe, it's also worth adding. As an option, the **Boundary conditions** section can be extented to **Initial & Boundary conditions**. Nonetherless, the current version is OK to me.

RemDelaporteMathurin commented:
@KulaginVladimir good shout! I added the initial conditions

RemDelaporteMathurin commented:
@KulaginVladimir I split ICs and BCs and added the initial conditions from XDMF files to the docs. Once it's compiled you should see it [here](https://festim--681.org.readthedocs.build/en/681/userguide/initial_conditions.html)

KulaginVladimir commented:
I think it is a nice guide. Would have been really helpful when I started using FESTIM 😄

jhdark commented on docs/source/userguide/beginners_guide.rst:
```diff
@@ -8,14 +8,29 @@ What do I need to know?
-
+New users should familiarise themselves with `Conda <https://anaconda.org/>`_.
 
+FESTIM is under version control with `git <https://git-scm.com/>`_. Even though users don't need git to install FESTIM, it is convenient to have a basic understanding of how it works. You can fin `git turorials <https://git-scm.com/doc>`_ to help you getting started. The `FESTIM source code <https://github.com/festim-dev/FESTIM>`_ is hosted on GitHub. It is recommended to sign up for free to GitH ...
```
```suggestion
FESTIM is under version control with `git <https://git-scm.com/>`_. Even though users don't need git to install FESTIM, it is convenient to understand how it works. You can find `git tutorials <https://git-scm.com/doc>`_ to help you getting started. The `FESTIM source code <https://github.com/festim-dev/FESTIM>`_ is hosted on GitHub. Signing up for free to GitHub is recommended to receive the latest updates, follow the development and submit `bug reports <https://github.com/festim-dev/FESTIM/issues/new/choose>`_.
```

jhdark commented on docs/source/userguide/beginners_guide.rst:
```diff
@@ -8,14 +8,29 @@ What do I need to know?
 Where can I find help?
 ----------------------
 
 If you're having an issue, the best way to find help is to join the `FESTIM discourse <https://festim.discourse.group>`_. Engage in discussions, ask questions, and connect with other users. This is a collaborative space where you can share your experiences and seek assistance.
```
```suggestion
If you're having an issue, the best way to find help is to join the `FESTIM discourse <https://festim.discourse.group>`_. Engage in discussions, ask questions, and connect with other users. This collaborative space allows you to share your experiences and seek assistance.
```

jhdark commented on docs/source/userguide/boundary_conditions.rst:
```diff
@@ -14,12 +14,13 @@ These BCs can be used for heat transfer or hydrogen transport simulations.
-   :class: tip
+.. note::
 
     Here, we set :code:`field=0` to specify this BC applies to the mobile hydrogen concentration. :code:`1` would stand for the trap 1 concentration and :code:`"T"` for temperature.
```
```suggestion
    Here, we set :code:`field=0` to specify that this BC applies to the mobile hydrogen concentration. :code:`1` would stand for the trap 1 concentration, and :code:`"T"` for temperature.
```

RemDelaporteMathurin commented:
> I think it is a nice guide. Would have been really helpful when I started using FESTIM 😄

Apologies for this... New users have a complete guide now! Thanks for your review!

jhdark commented on docs/source/userguide/boundary_conditions.rst:
```diff
@@ -81,7 +82,7 @@ Some BCs are specific to hydrogen transport. FESTIM provides a handful of conven
 ^^^^^^^^^^^^^^^^^^
 
-Recombination flux can be set on boundaries as: :math:`Kr \, c_\mathrm{m}^n`.
+A recombination flux can be set on boundaries as: :math:`Kr \, c_\mathrm{m}^n` (see :class:`festim.RecombinationFlux`).
```
```suggestion
A recombination flux can be set on boundaries as follows: :math:`Kr \, c_\mathrm{m}^n` (See :class:`festim.RecombinationFlux`).
```

jhdark commented on docs/source/userguide/boundary_conditions.rst:
```diff
@@ -127,28 +128,8 @@ Similarily, the mobile concentration can be set from Henry's law of solubility :
-
-.. admonition:: Note
-   :class: tip
+A plasma implantation can be approximated by a Dirichlet boundary condition with the class :class:`festim.ImplantationDirichlet` . Refer to the :ref:`theory` section for more details.
```
```suggestion
Plasma implantation can be approximated by a Dirichlet boundary condition with the class :class:`festim.ImplantationDirichlet` . Refer to the :ref:`theory` section for more details.
```

jhdark commented on docs/source/userguide/export_post_processing.rst:
```diff
@@ -1,3 +1,229 @@
+^^^^^^^^^^^
+
+The most straightforward way to export solutions (concentrations, temperature) with FESTIM is to use the :class:`festim.XDMFExport` class.
+This class leverages the ``XDMFFile`` class of ``fenics`` and allows to export solutions in the XDMF format.
```
```suggestion
This class leverages the ``XDMFFile`` class of ``fenics`` and allows solutions to be exported in the XDMF format.
```

jhdark commented on docs/source/userguide/export_post_processing.rst:
```diff
@@ -1,3 +1,229 @@
+The file can then be opened in Paraview or any other software that can read XDMF files.
+
+For transient simulations, by default, :class:`festim.XDMFExport` will export the solution at each timestep.
+It is however possible to change this behaviour in order to limit the number of times the file is written to.
```
```suggestion
It is possible to change this behaviour to limit the number of times the file is written to.
```

jhdark commented on docs/source/userguide/export_post_processing.rst:
```diff
@@ -1,3 +1,229 @@
+
+For transient simulations, by default, :class:`festim.XDMFExport` will export the solution at each timestep.
+It is however possible to change this behaviour in order to limit the number of times the file is written to.
+By setting the ``mode`` attribute to ``10`` for example, the solution will be exported every 10 timesteps.
```
```suggestion
By setting the ``mode`` attribute to ``10``, for example, the solution will be exported every 10 timesteps.
```

jhdark commented on docs/source/userguide/export_post_processing.rst:
```diff
@@ -1,3 +1,229 @@
+        )
+    ]
+
+In order for the XDMF file to be readable by FESTIM, the ``checkpoint`` attribute must be set to ``True``.
```
Do you mean paraveiw here?

```suggestion
The ``checkpoint`` attribute must be set to ``True`` for the XDMF file to be readable by Paraveiw.
```

RemDelaporteMathurin replied:
No I mean FESTIM

jhdark commented on docs/source/userguide/export_post_processing.rst:
```diff
@@ -1,3 +1,229 @@
+TXT export (1D)
+^^^^^^^^^^^^^^^
+
+The ``TXTExport`` class allows to export solutions in a simple text format.
```
```suggestion
The ``TXTExport`` class allows solutions to be exported in a simple text format.
```

jhdark commented on docs/source/userguide/export_post_processing.rst:
```diff
@@ -1,3 +1,229 @@
+    my_export = F.TXTExport(field="solute", filename="./mobile_conc.txt")
+
+Adding this export to the simulation object will produce a file called ``mobile_conc.txt`` in the current directory.
+This file will contain the solution of the ``solute`` field at degree of freedom of the mesh and at each timestep.
```
not sure about this one

```suggestion
This file will contain the solution of the ``solute`` field at the degree of freedom of the mesh and at each timestep.
```

jhdark commented on docs/source/userguide/export_post_processing.rst:
```diff
@@ -1,3 +1,229 @@
+^^^^^^^^^^^
+
+If information about the solution at a specific point is needed, the :class:`festim.PointValue` class can be used.
+It is implemented as a derived quantity, see :ref:`Derived quantities` for more information. Here are a few examples:
```
```suggestion
It is implemented as a derived quantity. See :ref:`Derived quantities` for more information. Here are a few examples:
```

jhdark commented on docs/source/userguide/export_post_processing.rst:
```diff
@@ -1,3 +1,229 @@
+In addition to exporting the actual solutions, it is possible to export derived quantities.
+For instance, you may want to compute the flux of mobile particles at a given boundary.
+
+First you want to create a :class:`festim.DerivedQuantities` object. This will encompass all the derived quantities you want to compute.
```
```suggestion
First, you want to create a :class:`festim.DerivedQuantities` object. This will encompass all the derived quantities you want to compute.
```

jhdark commented on docs/source/userguide/export_post_processing.rst:
```diff
@@ -1,3 +1,229 @@
+For instance, you may want to compute the flux of mobile particles at a given boundary.
+
+First you want to create a :class:`festim.DerivedQuantities` object. This will encompass all the derived quantities you want to compute.
+Then you can add the derived quantities you want to compute to this object.
```
```suggestion
Then, you can add the derived quantities you want to compute to this object.
```

jhdark commented on docs/source/userguide/initial_conditions.rst:
```diff
@@ -0,0 +1,44 @@
+        field=0
+    )
+
+In the snipset above, the initial condition is read from the file ``ic_file.xdmf``.
```
```suggestion
In the snippet above, the initial condition is read from the file ``ic_file.xdmf``.
```

jhdark commented on docs/source/userguide/initial_conditions.rst:
```diff
@@ -0,0 +1,44 @@
+
+.. note::
+
+    The XDMF file must be readable. To do so, the XDMF file has to be created with checkpointing on. See :class:`festim.XDMFExport`.
```
```suggestion
    The XDMF file must be readable. To do so, the XDMF file must be created with checkpointing. See :class:`festim.XDMFExport`.
```

jhdark commented on docs/source/userguide/materials.rst:
```diff
@@ -4,7 +4,7 @@ Materials
 
 Materials are key components of hydrogen transport simulations. They hold the properties like diffusivity, solubility and even thermal properties like thermal conductivity or heat capacity.
```
```suggestion
Materials are vital components of hydrogen transport simulations. They hold diffusivity, solubility and thermal properties like thermal conductivity or heat capacity.
```

jhdark commented on docs/source/userguide/materials.rst:
```diff
@@ -22,7 +22,7 @@ Materials are then assigned to the model:
-The :code:`Material` class has three required arguments:
+The :class:`festim.Material` class has three required arguments:
 
 * :code:`id`: a unique id given to the material/volume. It is useful when defining volumetric source terms or exports. Several ids can be given to the same material if multiple volumes have the same material.
```
```suggestion
* :code:`id`: a unique id given to the material/volume. It is useful when defining volumetric source terms or exports. Several id's can be given to the same material if multiple volumes have the same material.
```

jhdark commented on docs/source/userguide/materials.rst:
```diff
@@ -35,12 +35,12 @@ Some other parameters are optional and are only required for some types of simul
 Integration with HTM
 --------------------
 
 H-transport-materials (HTM) is a python database of hydrogen transport properties.
```
```suggestion
H-transport-materials (HTM) is a Python database of hydrogen transport properties.
```

jhdark commented on docs/source/userguide/sources.rst:
```diff
@@ -5,4 +5,62 @@ Sources
+Implantation flux
+-----------------
+
+Hydrogen implanted in a material can be simulated by a gaussian-shaped volumetric source with the :class:`festim.ImplantationFlux` class.
```
```suggestion
Hydrogen implanted in a material can be simulated by a Gaussian-shaped volumetric source with the :class:`festim.ImplantationFlux` class.
```

jhdark commented on docs/source/userguide/stepsize.rst:
```diff
@@ -0,0 +1,43 @@
+
+For transient problems, a :class:`festim.Stepsize` is required.
+It represents the time discretisation of the problem.
+Here is an example to create a stepsize of 1.2 seconds:
```
```suggestion
Here is an example creating a stepsize of 1.2 seconds:
```

jhdark commented on docs/source/userguide/stepsize.rst:
```diff
@@ -0,0 +1,43 @@
+    import festim as F
+    my_stepsize = F.Stepsize(initial_value=1.2)
+
+To make use of the adaptive time stepping implemented in FESTIM, the arguments ``stepsize_change_ratio`` needs to be set to a value above 1.
```
```suggestion
To use of the adaptive time stepping implemented in FESTIM, the arguments ``stepsize_change_ratio`` needs to be set to a value above 1.
```

jhdark commented on docs/source/userguide/stepsize.rst:
```diff
@@ -0,0 +1,43 @@
+It will shrink by the same ratio when the solver needs more than 4 iterations to converge.
+This is extremely useful when solving transient problems with a large time range, as the time step will be large when the solution is smooth and small when the solution is changing rapidly.
+Moreover, if the solver doesn't converge, the stepsize will be reduced and the solver will be called again.
+Setting the ``dt_min`` argument will prevent the stepsize to become too small and will stop the simulation when this happens.
```
```suggestion
Setting the ``dt_min`` argument will prevent the stepsize from becoming too small and will stop the simulation when this happens.
```

jhdark commented on docs/source/userguide/stepsize.rst:
```diff
@@ -0,0 +1,43 @@
+
+    my_stepsize = F.Stepsize(initial_value=1.2, stepsize_change_ratio=1.5, dt_min=1e-6, t_stop=10, stepsize_stop_max=1.5)
+
+To make sure the simulation passes through specific times, the ``milestones`` argument can be used.
```
```suggestion
The ``milestones`` argument can be used to make sure the simulation passes through specific times.
```

jhdark commented on docs/source/userguide/traps.rst:
```diff
@@ -92,9 +92,7 @@ Grouped-trap
 
 Let's imaging a case where you have two subdomains. Trap 1 is defined only in the first subdomain, whereas Trap 2 is defined in the second.
```
```suggestion
Let's imagine a case where you have two subdomains. Trap 1 is defined only in the first subdomain, whereas Trap 2 is defined in the second.
```

jhdark commented on docs/source/userguide/traps.rst:
```diff
@@ -92,9 +92,7 @@ Grouped-trap
 
 Let's imaging a case where you have two subdomains. Trap 1 is defined only in the first subdomain, whereas Trap 2 is defined in the second.
 It would be possible to simply define one trap in each subdomain.
```
```suggestion
It would be possible to define one trap in each subdomain.
```

jhdark commented on docs/source/userguide/traps.rst:
```diff
@@ -92,9 +92,7 @@ Grouped-trap
-
-
-But grouping traps together helps saving computational time.
+Grouping traps together helps saving computational time as it reduces the number of degrees of freedom.
```
```suggestion
Grouping traps together helps save computational time by reducing the number of degrees of freedom.
```

jhdark commented on docs/source/userguide/troubleshooting.rst:
```diff
@@ -2,18 +2,64 @@
--------------------
-Where to find help?
--------------------
+If you're having an issue, the best way to find help is to join the `FESTIM discourse <https://festim.discourse.group>`_. Engage in discussions, ask questions, and connect with other users. This is a collaborative space where you can share your experiences and seek assistance.
```
```suggestion
If you're having an issue, the best way to find help is to join the `FESTIM discourse <https://festim.discourse.group>`_. Engage in discussions, ask questions, and connect with other users. This collaborative space allows you to share your experiences and seek assistance.
```

jhdark commented on docs/source/userguide/troubleshooting.rst:
```diff
@@ -2,18 +2,64 @@
-------------------
-Convergence issues
-------------------
+Although FESTIM is designed to be easy to use, there are some common issues that users may encounter. This section provides some guidance on how to resolve these issues.
```
```suggestion
Although FESTIM is designed to be easy to use, users may encounter some common issues. This section provides some guidance on how to resolve these issues.
```

jhdark commented on docs/source/userguide/troubleshooting.rst:
```diff
@@ -2,18 +2,64 @@
-Convergence issues
-------------------
+Although FESTIM is designed to be easy to use, there are some common issues that users may encounter. This section provides some guidance on how to resolve these issues.
+At the end of the day, we are simply solving a set of equations using the finite element method, and are therefore subject to the same issues that other FEM codes face.
```
```suggestion
We are simply solving a set of equations using the finite element method, and are, therefore, subject to the same issues that other FEM codes face.
```

jhdark commented on docs/source/userguide/troubleshooting.rst:
```diff
@@ -2,18 +2,64 @@
+Solver doesn't converge
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The first thing to check are the details of the Newton solver iterations.
```
```suggestion
The first thing to check is the details of the Newton solver iterations.
```

jhdark commented on docs/source/userguide/troubleshooting.rst:
```diff
@@ -2,18 +2,64 @@
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The first thing to check are the details of the Newton solver iterations.
+To do so, you need to set the ``log_level`` to ``20`` (default is ``40``).
```
```suggestion
To do so, you must set the ``log_level`` to ``20`` (default is ``40``).
```

jhdark commented on docs/source/userguide/troubleshooting.rst:
```diff
@@ -2,18 +2,64 @@
+
+From there, depending on the behaviour of the solver, you can try the following:
+- if the solver diverges, try reducing the time step and/or mesh refinement. It is often
+ useful to inspect the fields to see if there are any obvious issues (like lack of refinement).
```
```suggestion
 helpful to inspect the fields to see if there are any obvious issues (like lack of refinement).
```

jhdark commented on docs/source/userguide/troubleshooting.rst:
```diff
@@ -2,18 +2,64 @@
+From there, depending on the behaviour of the solver, you can try the following:
+- if the solver diverges, try reducing the time step and/or mesh refinement. It is often
+ useful to inspect the fields to see if there are any obvious issues (like lack of refinement).
+- if the solver converges to a value above the tolerance, try increasing the tolerance.
```
```suggestion
- If the solver converges to a value above the tolerance, try increasing the tolerance.
```

jhdark reviewed (changes requested):
Just some minor typos and grammar from me, although if I ever find a few days free I could go through all the docs more thoroughly. Otherwise thanks a lot for doing this! Its tedious work but hugely beneficial to new users!

RemDelaporteMathurin commented:
@jhdark thanks for your feedback. I pushed the suggested changes, we are in no rush to merge this so if you need more time no worries.

jhdark reviewed (approved):
Looks good!
